keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,980 @@
+"prefer node-local, failover to same zone, then route anywhere" While this
+specific behavior is a non-goal for the initial implementation of the
+`routingPreference` field, the design allows for the potential introduction of
+such a preference in future refinements.
+
+## Goals
+
+* **Enhanced User Control:** Provide users with a knob to specify a preference that
+  MAY influence the traffic routing of a Kubernetes Service.
+
+* **Guidance for Implementations:** Offer standard routing preference values
+  that implementations can choose to recognize and support, promoting a degree
+  of commonality.
+
+* **Flexibility:** Allow implementations to interpret and support standard
+  heuristics in a manner that aligns with their capabilities and design.
+
+* **Extensibility:** Enable Kubernetes implementations to introduce and support
+  innovative routing strategies (for example, those based on topology, latency,
+  or other custom heuristics) that can evolve alongside users needs and
+  infrastructure capabilities.
+
+## Non-Goals
+
+* **Strict Routing Guarantees:** The `routingPreference` field will not
+  enforce deterministic routing paths. It serves as a mechanism for expressing
+  hints and preferences that implementations can consider when making routing
+  decisions.
+
+* **Replacement of Traffic Policies:** The new field is complementary to
+  `internalTrafficPolicy` and `externalTrafficPolicy`. It does not aim to
+  substitute their role in enforcing strict traffic locality.
+
+* **Immediate Support for All Possible Heuristics:** The initial implementation
+  focuses on a core set of heuristics. Addition of new heuristics (like
+  `Local` for Node local preference) could be explored in future
+  refinements.
+
+## Proposal
+
+Add a new field, `routingPreference`, to the Service specification. This field
+will serve as preference or hint for the underlying implementation to consider
+while making routing decisions. It does not offer strict routing guarantees.
+
+The field will support the following initial values:
+
+
+* `Default`: Indicates no specific routing preference. The user delegates the
+  routing decision to the implementation, allowing it to apply its best-effort
+  strategy.
+* `Close`: Indicates a preference for routing traffic to endpoints that
+  are topologically proximate to the client. The interpretation of
+  "topologically proximate" may vary across implementations and could encompass
+  endpoints within the same node, rack, zone, or even region.
+
+Implementations SHOULD support the standard values. While some flexibility in
+interpretation is permitted, implementations should aim to align their behavior
+with the described intent of these preferences as closely as possible.
+
+Implementations may support additional routing heuristics using values of the
+form `<domain>/<heuristicName>`. Heuristics without a domain prefix will be
+reserved for potential future standardization.
```

Comment:
I'd personally recommend leaving this as an optional future enhancement to limit the scope of this KEP for now. We can always loosen validation to allow this in the future if there's demand for it.